rmw_zenoh_cpp/src/detail/rmw_subscription_data.cpp

```diff
@@ -240,19 +196,51 @@ bool SubscriptionData::init()
     // We set consolidation to none as we need to receive transient local messages
     // from a number of publishers. Eg: To receive TF data published over /tf_static
     // by various publishers.
-    sub_options.query_consolidation = z_query_consolidation_none();
-    ze_owned_querying_subscriber_t sub;
-    if (ze_declare_querying_subscriber(
-        sess_->loan(), &sub, z_loan(sub_ke), z_move(callback), &sub_options))
-    {
+    sub_options.query_consolidation =
+      zenoh::QueryConsolidation(zenoh::ConsolidationMode::Z_CONSOLIDATION_MODE_NONE);
+
+    std::weak_ptr<SubscriptionData> data_wp = shared_from_this();
+    auto sub = context_impl->session()->declare_querying_subscriber(
+      sub_ke,
+      [data_wp](const zenoh::Sample & sample) {
+        auto sub_data = data_wp.lock();
+        if (sub_data == nullptr) {
+          RMW_ZENOH_LOG_ERROR_NAMED(
+            "rmw_zenoh_cpp",
+            "Unable to obtain SubscriptionData from data for %s.",
+            sample.get_keyexpr().as_string_view());
+          return;
+        }
+
+        auto attachment = sample.get_attachment();
+        if (!attachment.has_value()) {
+          RMW_ZENOH_LOG_ERROR_NAMED(
+            "rmw_zenoh_cpp",
+            "Unable to obtain attachment")
+          return;
+        }
+
+        auto attachment_value = attachment.value();
+        AttachmentData attachment_data(attachment_value);
+
+        sub_data->add_new_message(
+          std::make_unique<SubscriptionData::Message>(
+            sample.get_payload().as_vector(),
```

Comment:
Is this safe to do?  The sample is a `constref` here, and we are basically taking a pointer.  Will that data still exist when we go to dereference it during `take_one_message` later?  How are we guaranteed that?

Reply:
The function name `as_vector` may not be appropriate... `as_vector` copies the data and then we move `std::vector` into the `Message`.

Reply:
I see, OK.  I guess what `as_vector` does is to copy the data from the internal list of bytes to a vector, and return it.  So the vector will outlive the Sample.  In that case, this is fine.